### PR TITLE
Allow crafting using non-empty components

### LINKED ISSUE #52969: Loaded items don't don't qualify as crafting components
### Describe the bug

Let's assume I wanted to craft a tailor's kit. I know I have everything I need, but for some reason it says I lack a sewing kit. So I unloaded the thread from my sewing kit, and it worked! To check it, I loaded my curved needle, and it also turned out red. Is it a bug, or a feature?

### Steps To Reproduce

1. Reload a sewing kit or curved needle with thread
2. Try crafting a tailor's kit
3. You can see you don't have the required item, even if you do (only loaded)

### Expected behavior

The crafting should work, also the leftover thread (or another "charge") should obviously not be destroyed.

### Versions and configuration

CDDA-tiles version 0.F-2, Debian Linux x86_64

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dseguin**
#### Summary
None

#### Purpose of change
Fixes #52969.
Currently, items can only be used as crafting ingredients when they are empty (no contents). This causes some extra work to unload the item before using it in the crafting recipe.

#### Describe the solution
There are 2 key steps to this fix:
1. Allow non-empty items to be used as crafting ingredients.
2. Unload the contents of each ingredient back into the player's inventory when consumed, so the contents are preserved.

#### Describe alternatives you've considered
It's possible to limit this feature to tool components only, seeing as those are mostly the source of the issue. Although there's nothing that prevents this working for non-tool containers.

#### Testing
<hr>
<details>
<summary><b><i>(Expand) Crafting a recipe using a container with charges as an item component</i></b></summary>

Step 1: add the ingredients for the Tailor's Kit to the inventory (Note: sewing kit and curved needle have contents)
inv_before

Step 2: view the recipe for the Tailor's Kit (Note: recipe is craftable)
can_craft

Step 3: craft the recipe
msg_log

Step 4: see that the contents of the consumed ingredients are in the inventory (400 thread)
inv_after
</details>
<hr>
<details>
<summary><b><i>(Expand) Crafting a recipe using a liquid-containing ingredient</i></b></summary>

Step 1: add ingredients for blood draw kit to inventory (Note: non-empty small plastic bottle)
inv_before

Step 2: view recipe for blood draw kit (Note: recipe is craftable)
craft_menu

Step 3: player is warned that one of the components contains a liquid and is prompted to proceed with the craft
container_prompt

Step 4: player is prompted to unload liquid from container
pour_liquid

Step 5: crafted blood draw kit
msg_log

Step 6: see the unloaded liquid in other selected container in inventory
inv_after
</details>
<hr>
<details>
<summary><b><i>(Expand) Crafting a recipe where a tool is also a component</i></b></summary>

Step 1: Re-add the old welding rig recipe (to use as an example)
```json
{
  "type": "recipe",
  "result": "weldrig",
  "category": "CC_ELECTRONIC",
  "subcategory": "CSC_ELECTRONIC_PARTS",
  "skill_used": "mechanics",
  "skills_required": [ "electronics", 4 ],
  "difficulty": 4,
  "time": 50000,
  "reversible": true,
  "decomp_learn": 4,
  "book_learn": [ [ "textbook_mechanics", 6 ], [ "textbook_electronics", 8 ], [ "textbook_fabrication", 8 ], [ "welding_book", 5 ] ],
  "using": [ [ "welding_standard", 10 ] ],
  "qualities": [ { "id": "HAMMER", "level": 2 }, { "id": "SAW_M", "level": 1 }, { "id": "WRENCH", "level": 1 } ],
  "components": [ [ [ "frame", 1 ] ], [ [ "soldering_iron", 1 ] ], [ [ "welder", 1 ], [ "welder_crude", 2 ] ], [ [ "cable", 5 ] ] ]
}
```

Step 2: add ingredients and tools for crafting the welding rig to inventory (Note: arc welder with charges, soldering iron with charges)
inv_before2

Step 3: crafting the welding rig consumes the arc welder loaded with the battery
craft_warn2

Step 4: see that the batteries for the consumed ingredients (soldering iron, arc welder) are in the inventory
inv_after2

Step 5: reload the remaining arc welder with the automatically unloaded battery and continue the craft
msg_log
</details>
<hr>

#### Additional context
Most of the changes are for making it possible to craft with liquid-containing ingredients.

### CODE REVIEW

**Comment by PatrikLundell** [comment]
I can see some potential issues:
- Usage of a loaded item rather than the unloaded one you intended to use, resulting in confusion when you (or your companions) fail to use the loaded item you had prepared. I believe there were issues in the past where the construction of a vehicle arc welder failed because the game decided to use the loaded welder as a component, and thus couldn't weld it together as the unloaded welder left behind couldn't weld.
- Usage of an item loaded with liquid. That can be messy... The same thing goes for using a container containing liquid (or, for that matter, "items" that can't be placed in an inventory, such as a dog in a pet pack (although there probably aren't any recipes for these items currently).

**Comment by dseguin** [comment]
Cheers for the information. I made some changes based on these situations and tested these each case:

<details>
<summary>(Expand) Crafting a recipe using a liquid-containing ingredient</summary>

Step 1: add ingredients for blood draw kit to inventory (Note: non-empty small plastic bottle)
inv_before

Step 2: view recipe for blood draw kit (Note: recipe is craftable)
craft_menu

Step 3: player is prompted to unload liquid from container
pour_liquid

Step 4: crafted blood draw kit
msg_log

Step 5: see the unloaded liquid in other selected container in inventory
inv_after
</details>

<details>
<summary>(Expand) Crafting a recipe where a tool is also a component</summary>

Step 1: Re-add the old welding rig recipe (to use as an example)
```json
{
  "type": "recipe",
  "result": "weldrig",
  "category": "CC_ELECTRONIC",
  "subcategory": "CSC_ELECTRONIC_PARTS",
  "skill_used": "mechanics",
  "skills_required": [ "electronics", 4 ],
  "difficulty": 4,
  "time": 50000,
  "reversible": true,
  "decomp_learn": 4,
  "book_learn": [ [ "textbook_mechanics", 6 ], [ "textbook_electronics", 8 ], [ "textbook_fabrication", 8 ], [ "welding_book", 5 ] ],
  "using": [ [ "welding_standard", 10 ] ],
  "qualities": [ { "id": "HAMMER", "level": 2 }, { "id": "SAW_M", "level": 1 }, { "id": "WRENCH", "level": 1 } ],
  "components": [ [ [ "frame", 1 ] ], [ [ "soldering_iron", 1 ] ], [ [ "welder", 1 ], [ "welder_crude", 2 ] ], [ [ "cable", 5 ] ] ]
}
```

Step 2: add ingredients and tools for crafting the welding rig to inventory (Note: arc welder with charges, soldering iron with charges)
inv_before2

Step 3: crafting the welding rig consumes the arc welder loaded with the battery
craft_warn2

Step 4: see that the batteries for the consumed ingredients (soldering iron, arc welder) are in the inventory
inv_after2

Step 5: reload the remaining arc welder with the automatically unloaded battery and continue the craft
msg_log
</details>

In a nutshell:
- Crafting with ingredients containing liquids behaves as you would expect (prompt to unload the liquid)
- Crafting with ingredients used as both tools and components has the same issue as before, but the charges are properly unloaded (I guess these recipes were changed for good reason, see #30009)

**Comment by Amoebka** [comment]
What happens if a player attempts to craft with a liquid container, and when prompted to select a new container cancels out with esc?

**Comment by dseguin** [comment]
> What happens if a player attempts to craft with a liquid container, and when prompted to select a new container cancels out with esc?

The liquid will be consumed with the container.

**Comment by Amoebka** [comment]
Well that doesn't seem like the desired/expected behaviour. Ideally the craft should get cancelled. Otherwise you end up with situations where players start crafting with a full container, realise they don't have a spare to pour the liquid into, and lose the liquid without being able to cancel.

**Comment by dseguin** [comment]
> Well that doesn't seem like the desired/expected behaviour. Ideally the craft should get cancelled. Otherwise you end up with situations where players start crafting with a full container, realise they don't have a spare to pour the liquid into, and lose the liquid without being able to cancel.

An important issue is that once the player presses the CONFIRM key on the selected recipe, they are passed the point-of-no-return. The point where items are consumed for the craft are too far in to revert (well, it would require a lot of refactoring).

What I can do is run a check for non-empty containers used as ingredients and prompt the player before proceeding with the craft:
container_prompt

Answering "No" cancels the craft and preserves the ingredients. Is this an acceptable solution?

**Comment by Amoebka** [comment]
Assuming it always tries to craft with empty bottles first and only prompts when there aren't any, yes, that's good. Perhaps it could check for suitable spare containers and print the same message as crafting liquids without a storage ("You have nowhere to store clean water, craft anyway" or something, already in the game)?
So if I have an empty small bottle the craft uses it first without asking.
If I only have full small plastic bottles, but have enough other containers to move the liquid to, the craft starts and asks me to choose new container(s).
If I only have full small plastic bottles and not enough other containers to move the liquid, the game forces a confirmation and destroys the liquid.

**Comment by dseguin** [comment]
> Assuming it always tries to craft with empty bottles first and only prompts when there aren't any, yes, that's good. Perhaps it could check for suitable spare containers and print the same message as crafting liquids without a storage ("You have nowhere to store clean water, craft anyway" or something, already in the game)? So if I have an empty small bottle the craft uses it first without asking. If I only have full small plastic bottles, but have enough other containers to move the liquid to, the craft starts and asks me to choose new container(s). If I only have full small plastic bottles and not enough other containers to move the liquid, the game forces a confirmation and destroys the liquid.

After quite a bit of work on this, I don't think that is going to happen for a few reasons:
- This requires 2 passes when processing items, first for empty items and second for non-empty items. The `use_amount` functions are already quite heavy, the second pass makes it pretty slow. This is just the nature of how these functions are designed.
- The item search functions for `map` and `Character` will need refactoring to properly track items used between passes. Those functions nest pretty deeply and span a large area of the codebase. It's very possible that I would introduce more bugs than I would be fixing here.
- With the current solution, the player is warned if they would use a non-empty liquid container. They can cancel the craft and empty the container on their own before trying the craft again.

**Comment by rumly111** [comment]
nice, thanks :)
